ccl/multiregionccl: TestMultiRegionTenantRegions failed

ccl/multiregionccl.TestMultiRegionTenantRegions [failed](https://mesolite.cluster.engflow.com/invocations/default/20c265c1-93c5-416f-a26b-8691eb62a621?testReportRun=6&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on release-25.3.5-rc @ [adc4b2ff64df47187d363c76db80a5ed4200aa35](https://github.com/cockroachdb/cockroach/commits/adc4b2ff64df47187d363c76db80a5ed4200aa35):


```
=== RUN   TestMultiRegionTenantRegions
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestMultiRegionTenantRegions1031793325
    test_log_scope.go:76: use -show-logs to present logs inline
    multiregion_system_table_test.go:434: error executing query="ALTER DATABASE system DROP REGION \"us-east2\"" args=[]: pq: failed to count leases
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestMultiRegionTenantRegions1031793325
--- FAIL: TestMultiRegionTenantRegions (25.90s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #154453 ccl/multiregionccl: TestMultiRegionTenantRegions failed [failed to count leases] [C-test-failure O-robot P-2 T-sql-foundations branch-release-25.4 no-test-failure-activity]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMultiRegionTenantRegions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-56375

--- a/pkg/sql/catalog/lease/BUILD.bazel
+++ b/pkg/sql/catalog/lease/BUILD.bazel
@@ -77,6 +77,7 @@ go_test(
     name = "lease_test",
     size = "large",
     srcs = [
+        "count_test.go",
         "helpers_test.go",
         "ie_writer_test.go",
         "kv_writer_test.go",

--- a/pkg/sql/catalog/lease/count.go
+++ b/pkg/sql/catalog/lease/count.go
@@ -208,8 +208,12 @@ func countLeasesByRegion(
 		} else {
 			err = queryRegionRows(ctx)
 		}
-		if err := handleRegionLivenessErrors(ctx, prober, region, err); err != nil {
+		skipRegion, err := handleRegionLivenessErrors(ctx, prober, region, err)
+		if err != nil {
 			return err
+		}
+		if skipRegion {
+			return nil
 		}
 		if values == nil {
 			return errors.New("failed to count leases")
@@ -231,27 +235,27 @@ func getCountLeaseColumns() string {
 }
 
 // handleRegionLivenessErrors handles errors that are linked to region liveness
-// timeouts.
+// timeouts. Return true if the region should be skipped.
 func handleRegionLivenessErrors(
 	ctx context.Context, prober regionliveness.Prober, region string, err error,
-) error {
+) (bool, error) {
 	if err != nil {
 		if regionliveness.IsQueryTimeoutErr(err) {
 			// Probe and mark the region potentially.
 			probeErr := prober.ProbeLiveness(ctx, region)
 			if probeErr != nil {
 				err = errors.WithSecondaryError(err, probeErr)
-				return err
+				return false, err
 			}
-			return errors.Wrapf(err, "count-lease timed out reading from a region")
+			return false, errors.Wrapf(err, "count-lease timed out reading from a region")
 		} else if regionliveness.IsMissingRegionEnumErr(err) {
 			// Skip this region because we were unable to find region in
 			// type descriptor. Since the database regions are cached, they
 			// may be stale and have dropped regions.
 			log.Dev.Infof(ctx, "count-lease skipping region %s due to error: %v", region, err)
-			return nil
+			return true, nil
 		}
-		return err
+		return false, err
 	}
-	return err
+	return false, err
 }

--- a/pkg/sql/catalog/lease/count_test.go
+++ b/pkg/sql/catalog/lease/count_test.go
@@ -1,0 +1,26 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package lease
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/stretchr/testify/require"
+)
+
+func TestHandleRegionLivenessErrorsSkipsMissingRegion(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	skip, err := handleRegionLivenessErrors(ctx, nil /* prober */, "us-east2", types.EnumValueNotYetPublicError)
+	require.NoError(t, err)
+	require.True(t, skip)
+}

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -425,7 +425,11 @@ func (m *Manager) WaitForInitialVersion(
 						sessionIDs, err = getSessionsHoldingDescriptor(ctx, txn, schemaID, region)
 					}
 					if err != nil {
-						return handleRegionLivenessErrors(ctx, prober, region, err)
+						skipRegion, handledErr := handleRegionLivenessErrors(ctx, prober, region, err)
+						if skipRegion {
+							return nil
+						}
+						return handledErr
 					}
 					sessionsPerRegion[region] = sessionIDs
 					expectedSessions += len(sessionIDs)
@@ -469,8 +473,12 @@ func (m *Manager) WaitForInitialVersion(
 						} else {
 							regionCount, err = countDescriptorsHeldBySessionIDs(ctx, txn, descIDsForSchema, region, sessionIDs)
 						}
-						if err := handleRegionLivenessErrors(ctx, prober, region, err); err != nil {
+						skipRegion, err := handleRegionLivenessErrors(ctx, prober, region, err)
+						if err != nil {
 							return err
+						}
+						if skipRegion {
+							return nil
 						}
 						count += regionCount
 						return nil
@@ -620,7 +628,13 @@ func (m *Manager) WaitForNewVersion(
 					regionStaleSessionCount, err = countSessionsHoldingStaleDescriptor(ctx, txn, desc, region)
 				}
 				if err != nil {
-					return handleRegionLivenessErrors(ctx, prober, region, err)
+					skipRegion, handledErr := handleRegionLivenessErrors(ctx, prober, region, err)
+					if handledErr != nil {
+						return handledErr
+					}
+					if skipRegion {
+						continue
+					}
 				}
 
 				staleSessionCount += regionStaleSessionCount

